### ISSUE TITLE
Help needed to create a Dockerfile

### ISSUE THREAD

**Issue by george-andrei** [r-follador/CubeTrek#1]
I have a bit of experience with docker, but nothing with java. If someone is willing to help me going through a full installation/deployment, I can probably help creating the Dockerfile.

PS1: I really like this project, looks really good! I hope only but success :) 

PS2: Maybe you should also consider GitHub sponsor?

**Comment by r-follador** [comment]
It might be helpful for setting up a dev environment. For a functional stand-alone version, I suggest that I'll create a slimmed down version with less dependencies on external services (such as removing email verification, no integration of Garmin Connect and Polar etc.). One of the problems is that all height data (HGT files) for the full planet will take up approx. 1TB, that makes self-hosting rather complicated...

**Comment by george-andrei** [comment]
That would be a good start. And you are right, 1TB of data can be much for a lot of people, but not everyone. An option would be to generate the height data per region. 
How does the application react if not all height data is present? Do you load that data only when generating a map? 
The generation and preparation of data could be done in a separate container than when the front-end would be. I already see you have different services, these can be self-contained. I am not a huge fan of 1 big container.

**Comment by srwareham** [comment]
Seconding easier install + dono! I think having a clean docker compose file to self host this would really help get more attention & users. I'd personally love to run this stack on my own metal. After all, data ownership & privacy are the features Strava can't touch comparatively. I think the features and finish here are quite brilliant.

**Comment by pto199** [comment]
Just came across this.  Looks great, but I need a docker to run this.  +1 for docker support.  This would be popular software with it

**Comment by pmffromspace** [comment]
I created a pull request with a first version🎉🚀, feel free to optimise, but I think its a working start..

https://github.com/r-follador/CubeTrek/pull/6/commits/3d55a3226597532e036d02ee3f4667fde4f2973c

**Comment by r-follador** [comment]
If anyone can support @pmffromspace in getting the dockerized version further on its track, that would be awesome

**Comment by george-andrei** [comment]
I will try to run this container and will get back with feedback 🙂 thanks!

**Comment by srwareham** [comment]
I was able to leverage @pmffromspace 's great work and create a fork of his fork (https://github.com/srwareham/CubeTrek) that I think is relatively straightforward to get up and running. I had build issues with upstream master at its current state and even several recent ones so I've stuck with using this older known working state. Note that this by definition won't have newer features.

My goal for this fork was to have the easiest setup possible by using docker compose. To get up and running you need to
1. Install [docker](https://docs.docker.com/engine/install/) and [docker compose](https://docs.docker.com/compose/install/)
2. Get your own api keys for [maptiler](https://www.maptiler.com/) and optionally for [google maps](https://developers.google.com/maps) & [cesium ion](https://ion.cesium.com/signin/). Place these in the docker compose environment variables, for instance `MAPTILER_API_KEY=ABCdefhijk123`.
3. In the CubeTrek folder, create a new nested folder `mkdir -p cubetrek_data/maps/HGT`. Populate it with unzipped [3DEM files](https://viewfinderpanoramas.org/Coverage%20map%20viewfinderpanoramas_org3.htm) that correspond to where your tracks will be in the world (@george-andrei  from my testing you only need elevation data files for regions where your actual tracks occur). If you want to use [1DEM files](https://e4ftl01.cr.usgs.gov/MEASURES/SRTMGL1.003/?_ga=2.33740263.797586985.1703110939-19353304.1703110938) do  `mkdir -p cubetrek_data/maps/HGT_1DEM` and put them there. This is what it could look like with 3DEM

``` bash
$ tree cubetrek_data                                                                                                                     
cubetrek_data
└── maps
    └── HGT
        ├── N31W044.hgt
        ├── N31W045.hgt
        ├── N32W033.hgt
        ├── N32W035.hgt
```

4. From the toplevel CubeTrek folder, run `docker compose up` and access your CubeTrek instance at http://localhost:8080. If you end up changing anything in the underlying code, run `docker compose up --build` so that the image will be rebuilt.

Basically, all "essential" changes are done directly in the docker compose file. Potentially less relevant API keys/configurations can be managed in the `production.env` file. I have also taken a different approach from @pmffromspace  and just made changes directly inline (i.e., I haven't compartmentalized things in a separate `docker` folder).

I would love to get this fork patched to work with master but will wait for upstream to fix what I suspect is a master branch that doesn't successfully build. I would also happily accept PRs.

**Comment by george-andrei** [comment]
You did an amazing job @srwareham and @pmffromspace!

Unfortunately I'm really bad with time right now, so don't wait for my input. ☹️

**Comment by srwareham** [comment]
As an update I've actually managed to get a working dockerized version of the latest master. Had to fix some bugs that might be from me using different library versions than upstream somehow? One of them required some LLM black magic I didn't bother digging into.

I also redid the disabling of cloudflare turnstile and email registration.

Either way, if you want to try the current latest master CubeTrek using docker compose here's the latest: https://github.com/srwareham/CubeTrek/tree/dockerized2
